guardian/prosemirror-elements
# feat: [repeater PR 3] Handle ingress of values into repeater nodes

## What does this change?

Implements the next phase of #234 – handling the way data moves in and out of repeater nodes. See that PR for a description of the new structure of these nodes.

To achieve this, we need to add code to iterate over repeater field **values** and create nodes from them recursively.

The work to iterate over repeater field **nodes** and create values from them recursively is incomplete, and should be added in a subsequent PR – we need to consolidate the code that produces these values before changing it, as this operation currently happens in two places.

## How to test

Over half of this PR's line count is used by tests – either expanding existing ones or writing new ones. They should pass, and we should be convinced that they cover ingress and egress cases in enough depth to give us confidence.

We can also inspect the result of this work at runtime. I've added a repeater field to the demo image element. If you run the library with `yarn start` and add an element of this sort ('Add demo image'), you'll be able to see that the repeater nodes have been added to the document content, which should mirror the diagram in #234: 

<img width="593" alt="Screenshot 2022-04-13 at 10 14 41" src="https://user-images.githubusercontent.com/7767575/163143119-acd4fa93-bfb0-4c9d-ab49-3cefb2a4bbb2.png">

One thing to note – the code that produces repeater field values is incomplete, and as a result values from repeater fields will be incorrect. This requires a bit of a rethink into how we structure repeater nodes, and is addressed in #265.
